pkg/virt-controller/watch/vm_test.go

```diff
@@ -212,13 +212,23 @@ var _ = Describe("VirtualMachine", func() {
 		shouldExpectVMFinalizerAddition := func(vm *virtv1.VirtualMachine) {
 			patch := fmt.Sprintf(`[{ "op": "test", "path": "/metadata/finalizers", "value": null }, { "op": "replace", "path": "/metadata/finalizers", "value": ["%s"] }]`, virtv1.VirtualMachineControllerFinalizer)
 
-			vmInterface.EXPECT().Patch(context.Background(), vm.Name, types.JSONPatchType, []byte(patch), &metav1.PatchOptions{}).Return(vm, nil)
+			vmInterface.EXPECT().Patch(context.Background(), vm.Name, types.JSONPatchType, []byte(patch), &metav1.PatchOptions{}).DoAndReturn(
+				func(_, _, _, _, _ any, _ ...any) (*virtv1.VirtualMachine, error) {
+					vm := vm.DeepCopy()
+					vm.Finalizers = append(vm.Finalizers, virtv1.VirtualMachineControllerFinalizer)
+					return vm, nil
+				})
 		}
 
 		shouldExpectVMFinalizerRemoval := func(vm *virtv1.VirtualMachine) {
 			patch := fmt.Sprintf(`[{ "op": "test", "path": "/metadata/finalizers", "value": ["%s"] }, { "op": "replace", "path": "/metadata/finalizers", "value": [] }]`, virtv1.VirtualMachineControllerFinalizer)
 
-			vmInterface.EXPECT().Patch(context.Background(), vm.Name, types.JSONPatchType, []byte(patch), &metav1.PatchOptions{}).Return(vm, nil)
+			vmInterface.EXPECT().Patch(context.Background(), vm.Name, types.JSONPatchType, []byte(patch), &metav1.PatchOptions{}).DoAndReturn(
```

Comment:
Is this now needed because of the addition of sanityExecute?
Previously this function was doing the right thing as it was just expecting a specific operation and not an actual object change.

Reply:
Well, the change should be done by the client and the modified object should be returned. This is what happens if you are running the code against a real Kubernetes API server. Therefore, before it was wrong as we were returning a pointer to some VM, which on top of it was modified by the `sync`.

Reply:
This wasn't doing any modifications. It was only expecting a patch. Yes, it did return a VM object, but the same one it got.
But now this actually modifies the object, so I'm a bit confused. Why is it the right thing to do?
I could expect the actual controller code to modify the object and not the test code.
What am I missing?

Reply:
With the previous code, not only did the caller get back a pointer to a VM object that's actively used by the test code, but it also didn't include the requested patch.
Now we actually apply the patch (easy since we already know what it is) and return a pointer to an object that we don't rely on and that won't change over time.

Reply:
@jean-edouard I'm still confused... previous code only had to observe the patch - not to perform it. 
and now we are performing the patch while there was something else in the controller that had to apply the patch.

Reply:
If we don't apply it nothing will. It worked before because the caller doesn't verify the change happened (nor should it).
The most important part is the deepcopy, emptying the finalizers is just a bonus and may not matter today, but could in the future.
@xpivarc please proofread!

Reply:
@jean-edouard thanks!

Reply:
@jean-edouard +100.
The fact that we did nothing in the past is wrong, it does not reflect what happens if you run the code against real API. 

Note: This can be still wrong but it is at least slightly better. We need https://github.com/kubevirt/kubevirt/pull/11297 to be as close as we can to the real behavior.

Reply:
perhaps it's better to wait. 

> The fact that we did nothing in the past is wrong, it does not reflect what happens if you run the code against real API.

Mainly because it's not a real API, it's a unit test. 
All we needed to do was to observe the call being made.
I would expect complex flows to be tested in a functional test.

Reply:
> I would expect complex flows to be tested in a functional test.

If our "unit tests" really were unit tests (they're already much more than that) and if functional test were cheap (they're not) then I would 100% agree with you.
However, in the current state of things, I think we should always prefer unit tests when possible, no matter how complex.
Just my opinion, not really related to this PR.

Reply:
>How do we know that the fake client behaves the same way as a real client?
Do we really need to play a catchup game between the fake client and the real one?

This is standard tooling from Kubernetes. We are not going to play any catch up as long as we don't want to test specific scenarios/bugs.

>We are responsible for the code in our controller and this is what we should test in the unit tests (whatever they are).

Right but we should make sure we are testing realistic scenarios. For example, right now we don't. The patch should return an object that is only adjusted by the changes expressed in the request but until now we would return an object that contains every change or non (depending if we are looking at the code before or after fixing the cache corruption).

>A unit test or integration test will never be a substitute for a functional test that exercises the functionality altogether.
Paying "cheap" for a test will make us pay a much higher price later.

+100, we should not remove functional test unless they are duplicates.